### PR TITLE
PyTorch dataset tutorial

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @iindyk,

This PR adds PyTorch tutorial and shows how to use Torchvision dataset as a source with Map and Filter stages. 

----
📚 Documentation preview 📚: https://google-grain--867.org.readthedocs.build/

### CODE REVIEW

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.17.1
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
+# PyTorch dataset tutorial
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:15]: Is it possible to have a more descriptive title? Maybe something like "How to use a Torchvision dataset in the Grain pipeline"?

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
++++
+
+In this tutorial we're going to learn how to migrate from Torchvision to Grain pipeline for [Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) dataset.
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:19]:
```suggestion
In this tutorial we're going to learn how to migrate from the Torchvision to the Grain pipeline for the [Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) dataset.
```

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+# PyTorch imports
+from torchvision import datasets
+from torchvision.transforms import ToTensor, Lambda
+
+rng = np.random.default_rng(0)
+```
+
+## Loading dataset
+
+`FashionMNIST` function provides access to our dataset. There are 60k samples in the dataset, where each one of them is a Pillow image instance together with a label.
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:50]:
```suggestion
The Torchvision `FashionMNIST` function provides access to our dataset. There are 60k samples in the dataset, where each one of them is a Pillow image instance together with a label.
```

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+```{code-cell} ipython3
+fashion_mnist = datasets.FashionMNIST(root="data", train=True, download=True)
+print(fashion_mnist)
+print(fashion_mnist[0])
+fashion_mnist[0][0]
+```
+
+Here, for example, we have an ankle boot (label 9). First we should acknowledge that both sample and label requires a dedicated preprocessing:
+- Sample should be a PyTorch tensor
+- Label needs to use one-hot encoding, so `9` becomes `[0,0,0,0,0,0,0,0,0,1]`
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:61]:
```suggestion
In this example, we have an ankle boot (label 9). First we should acknowledge that both sample and label require dedicated preprocessing:
- The sample should be a PyTorch tensor;
- The label needs to use one-hot encoding, so `9` becomes `[0,0,0,0,0,0,0,0,0,1]`.
```

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+```{code-cell} ipython3
+nrows, ncols = 4, 8
+fig, axs = plt.subplots(nrows, ncols)
+for x in range(ncols):
+    for y in range(nrows):
+        axs[y, x].imshow(fashion_mnist.data[rng.integers(len(fashion_mnist))])
+        axs[y, x].set_axis_off()
+```
+
+`torchvision`'s function, which imports the dataset, already offers `transform` and `target_transform` arguments. Let's confirm that `ToTensor` and `Lambda` do what we anticipate.
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:74]:
```suggestion
`torchvision`'s `dataset` function, which imports the dataset, already offers `transform` and `target_transform` arguments. Let's confirm that `ToTensor` and `Lambda` do what we anticipate.
```

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        lambda y: torch.zeros(10, dtype=torch.float).scatter_(0, torch.tensor(y), value=1)
+    )
+)
+```
+
+```{code-cell} ipython3
+(fashion_mnist_2[0][0].size(), fashion_mnist_2[0][1])
+```
+
+The resulting contents are PyTorch tensor and one-hot encoded label indeed.
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:92]:
```suggestion
The resulting contents are a PyTorch tensor and a one-hot encoded label, as expected.
```

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+from PIL.Image import Image
+
+# PyTorch imports
+from torchvision import datasets
+from torchvision.transforms import ToTensor, Lambda
+
+rng = np.random.default_rng(0)
+```
+
+## Loading dataset
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:48]:
```suggestion
## Loading the dataset with Torchvision
```

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+)
+```
+
+```{code-cell} ipython3
+(fashion_mnist_2[0][0].size(), fashion_mnist_2[0][1])
+```
+
+The resulting contents are PyTorch tensor and one-hot encoded label indeed.
+
+Now let's move to Grain! Given the unprocessed dataset we want to arrive with the same transformation. You might have noticed in the "Dataset basics" tutorial that defining a data source requires implementing a class that inherits from `grain.sources.RandomAccessDataSource`. With PyTorch datasets we don't need it - the dataset is already compliant with the protocol.
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:94]: A suggestion but the main point is I think this should be a header; feel free to choose a section name that fits better.

```suggestion
## Processing the dataset with Grain

Now let's move to Grain! Given the unprocessed dataset we want to arrive with the same transformation. You might have noticed in the ["Dataset basics" tutorial](please add a link if possible) that defining a data source requires implementing a class that inherits from `grain.sources.RandomAccessDataSource`. With PyTorch datasets we don't need it - the dataset is already compliant with the protocol.
```

Comment (melissawm) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md:94]: It's not very clear to me what the sentence below means:
```
Given the unprocessed dataset we want to arrive with the same transformation.
```
Maybe reword to be less vague?

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+First let's define our data source with random access capabilities.
+
+```{code-cell} ipython3
+fashion_source = fashion_mnist
+```
+
+## Preprocessing
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:102]: If the previous suggestion is accepted, maybe this is not needed?

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+Next let's move to the preprocessing stage. We implement a custom class which inherits from `grain.transforms.Map` and implements a `map` method. In this method we instantiate a PyTorch tensor and devise a one-hot encoded label.
+
+To make this rewrite more compelling let's use Grain capabilities beyond simple samples traversal. Assuming that we are interested in shoes only (labels: `5` - sandals, `7` - sneakers, and `9` - ankle boots) we implement a `Filter` class with `filter` that allows us to discard unwanted samples.
```
Comment (melissawm) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:106]:
```suggestion
To make this rewrite more compelling, let's use Grain's capabilities beyond simple samples traversal. Assuming that we are interested in shoes only (labels: `5` - sandals, `7` - sneakers, and `9` - ankle boots) we implement a `Filter` class with a `filter` that allows us to discard unwanted samples.
```

**Review by melissawm — COMMENTED** [review]
Hi @mtsokol , left a couple of suggestions for you. Let me know if anything is not clear.

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+```{code-cell} ipython3
+class Preprocess(grain.transforms.Map):
```
Comment (iindyk) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:109]: ToTensorAndOneHot ?

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+    def map(self, element: tuple[Image, int]) -> tuple[torch.Tensor, torch.Tensor]:
+        data = self.to_tensor(element[0])
+        target = (
+            torch.zeros(10, dtype=torch.float).scatter_(0, torch.tensor(element[1]), value=1)
+        )
+        return (data, target)
+
+
+class Filter(grain.transforms.Filter):
```
Comment (iindyk) [inline docs/tutorials/data_sources/pytorch_dataset_tutorial.md:120]: KeepShoesOnly?

**Review by iindyk — COMMENTED** [review]
thanks! +1 to Melissa's comments and added few more

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 12 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 11 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+In this tutorial we're going to learn how to migrate from Torchvision to Grain pipeline for [Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) dataset.
+
++++
+
+## Setup
+
+Here we require `grain`, `torch`, and `torchvision` dependencies installed but the last will be used primarily for dataset access. Additionally we want `matplotlib` in our environment for displaying samples.
+
+```{code-cell} ipython3
+# @test {"output": "ignore"}
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Indeed it's some legacy code - removed!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 10 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 9 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: That's correct - done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 7 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 6 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 4 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

File: docs/tutorials/data_sources/pytorch_dataset_tutorial.md (hunk 2 above)
Comment (mtsokol) [reply docs/tutorials/data_sources/pytorch_dataset_tutorial.md]: Done!

**Comment by mtsokol** [comment]
@iindyk I included all comments from both reviews - it's ready from my side!

**Review by iindyk — APPROVED** [review]
